### PR TITLE
AWS: Experiment support for multiple AZs / subnets

### PR DESCRIPTION
**Pull request by justinsb**
As part of the Ubernetes Lite work, we need to be able to launch nodes in multiple AZs.  

(Ubernetes Lite = stepping-stone version of Ubernetes, which runs a single cluster in multiple AZs and has minimal changes to the scheduler so that it makes decisions that are reasonably zone-aware.)

This change allows kube-up to be run multiple times, sharing a master.  It could also be used to create a second ASG of nodes, for example if users wanted to mix spot & on-demand instances (even in a single AZ).

This change is only kube-up work; though users can launch nodes in multiple AZs, the k8s scheduler is totally unaware of zones.  My plan here is to do this PR for AWS, then tag the instances with the AZ and make the scheduler do AZ-spreading in an RC, then tag volumes with AZs and do AZ-constraints for pods with volumes.  Then to do the same work for GCE.  I believe that will be the entire extent of Ubernetes-Lite; more advanced functionality should be focused on Ubernetes-Full.

To use:

```
export KUBE_PROVIDER=aws
KUBE_AWS_ZONE=us-west-2a cluster/kube-up.sh
KUBE_AWS_ZONE=us-west-2b KUBE_SUBNET_CIDR=172.20.1.0/24 KUBE_SHARE_MASTER=true cluster/kube-up.sh
```

cc @quinton-hoole @brendandburns

### CODE REVIEW

File: cluster/aws/util.sh [hunk 1]
```diff
@@ -16,6 +16,19 @@
 
 # A library of helper functions and constant for the local config.
 
+# Experimental flags can be removed/renamed at any time.
+# The intent is to allow experimentation/advanced functionality before we
+# are ready to commit to supporting it.
+# Experimental functionality:
+#   KUBE_SHARE_MASTER=true
+#     Detects an existing master and reuse it; useful if you want to
```
Comment (quinton-hoole) [inline cluster/aws/util.sh]: nit: ("Detects" and "reuses" or "Detect and reuse")

File: cluster/aws/util.sh [hunk 2]
```diff
@@ -16,6 +16,19 @@
...
+#     create more nodes, perhaps with a different instance type or in
+#     a different subnet/AZ
+#   KUBE_SUBNET_CIDR=172.20.1.0/24
```
Comment (quinton-hoole) [inline cluster/aws/util.sh]: nit: Do we want to make this comment consistent with the default?

File: cluster/aws/util.sh [hunk 3]
```diff
@@ -25,7 +38,9 @@ source "${KUBE_ROOT}/cluster/common.sh"
 ALLOCATE_NODE_CIDRS=true
 
 NODE_INSTANCE_PREFIX="${INSTANCE_PREFIX}-minion"
-ASG_NAME="${NODE_INSTANCE_PREFIX}-group"
+
+# The ASG name must be unique, so we include the zone
```
Comment (quinton-hoole) [inline cluster/aws/util.sh]: nit: Expand this acronym to "Auto Scaling Group", at least once, upon introduction.

**Comment by quinton-hoole** [comment]
@justinsb very minor, verging on pedantic, nits.  LGTM.

File: cluster/aws/util.sh (hunk 2 above)
Comment (justinsb) [inline cluster/aws/util.sh]: I was trying to provide a hint as to the typical usage - I imagine most people will not specify the default, but may then need a hint for the second AZ.  (Same as KUBE_SHARE_MASTER=true)

File: cluster/aws/util.sh (hunk 3 above)
Comment (justinsb) [inline cluster/aws/util.sh]: Pushed fix in latest commit

File: cluster/aws/util.sh (hunk 1 above)
Comment (justinsb) [inline cluster/aws/util.sh]: Pushed fix in latest commit

**Comment by quinton-hoole** [comment]
LGTM

**Comment by quinton-hoole** [comment]
e2e failure was a flake on cluster initialization, unrelated to this PR. 

@k8s-bot test this again please

**Comment by quinton-hoole** [comment]
@k8s-bot unit test this please

**Comment by quinton-hoole** [comment]
@k8s-bot unit test this
I was too polite last time, apparently :-)

**Comment by quinton-hoole** [comment]
@k8s-bot unit test this

@jlowdermilk tells me that my previous attempts failed because he was in the process of rebooting the test infrastructure.

**Comment by druidsbane** [comment]
There is a problem running this using the 1.2alpha7 release.  Am I doing something wrong?  I'm following the instructions up above exactly.  I think the only difference is hardcoding a master IP using MASTER_RESERVED_IP and modifying the number of nodes to 2 instead of the default of 4.  I'm also using the West region as in the example up top.

When running I get the following errors:

```
Using Route Table rtb-xxxxxxxx
/Users/username/kubernetes/cluster/../cluster/../cluster/aws/../../cluster/common.sh: line 420: @: unbound variable
/Users/username/kubernetes/cluster/../cluster/../cluster/aws/../../cluster/common.sh: line 420: @: unbound variable
/Users/username/kubernetes/cluster/../cluster/../cluster/aws/../../cluster/common.sh: line 420: @: unbound variable
/Users/username/kubernetes/cluster/../cluster/../cluster/aws/../../cluster/common.sh: line 420: @: unbound variable
/Users/username/kubernetes/cluster/../cluster/../cluster/aws/../../cluster/common.sh: line 420: @: unbound variable
/Users/username/kubernetes/cluster/../cluster/../cluster/aws/../../cluster/common.sh: line 420: @: unbound variable
/Users/username/kubernetes/cluster/../cluster/../cluster/aws/../../cluster/common.sh: line 420: @: unbound variable
/Users/username/kubernetes/cluster/../cluster/../cluster/aws/../../cluster/common.sh: line 420: @: unbound variable
/Users/username/kubernetes/cluster/../cluster/../cluster/aws/../../cluster/common.sh: line 420: @: unbound variable
/Users/username/kubernetes/cluster/../cluster/../cluster/aws/../../cluster/common.sh: line 549: KUBE_BEARER_TOKEN: unbound variable
/Users/username/kubernetes/cluster/../cluster/../cluster/aws/../../cluster/common.sh: line 420: @: unbound variable
/Users/username/kubernetes/cluster/../cluster/../cluster/aws/../../cluster/common.sh: line 420: @: unbound variable
/Users/username/kubernetes/cluster/../cluster/../cluster/aws/../../cluster/common.sh: line 420: @: unbound variable
/Users/username/kubernetes/cluster/../cluster/../cluster/aws/../../cluster/common.sh: line 420: @: unbound variable
/Users/username/kubernetes/cluster/../cluster/../cluster/aws/../../cluster/common.sh: line 420: @: unbound variable
/Users/username/kubernetes/cluster/../cluster/../cluster/aws/../../cluster/common.sh: line 420: @: unbound variable
/Users/username/kubernetes/cluster/../cluster/../cluster/aws/../../cluster/common.sh: line 420: @: unbound variable
Starting Master

A client error (InvalidParameterValue) occurred when calling the RunInstances operation: Address 172.20.0.9 does not fall within the subnet's address range
```

**Comment by justinsb** [comment]
@druidsbane thanks for trying it.  I think you found a bug with the error/messages you posted.  But also we changed the syntax here... can you confirm you are using `export KUBE_USE_EXISTING_MASTER=true`. 

There's no reason that you would know this unless you followed all the PRs.  Which reminds me that docs need to be fixed!

**Comment by druidsbane** [comment]
@justinsb thanks for that!  Apparently that was it.  I have another error right now but it is likely unrelated to this:

```
Could not detect Kubernetes master node IP.  Make sure you've launched a cluster with 'kube-up.sh'
```

I suspect that's because I can't connect to my master.  I'll dig in and report back here if things still fail once I can connect to my master again.  It is using a fixed master IP so maybe something broke in the AWS setup for 1.2alpha.

Thanks again for your help!

**Comment by druidsbane** [comment]
@justinsb Unfortunately I still hit this error:
```
Could not detect Kubernetes master node IP.  Make sure you've launched a cluster with 'kube-up.sh'
```

Despite the cluster being up in one AZ and being able to connect to it the script to bring up the second AZ is failing.  Any thoughts?

```
export KUBERNETES_PROVIDER=aws
export MASTER_RESERVED_IP=<someIP>
export NODE_SIZE=m4.large
export MASTER_SIZE=m4.large

# this part succeeds
KUBE_AWS_ZONE=us-west-2a kube-up.sh

# this command fails with: Could not detect Kubernetes master node IP.  Make sure you've launched a cluster with 'kube-up.sh'
KUBE_AWS_ZONE=us-west-2b KUBE_SUBNET_CIDR=172.20.1.0/24 KUBE_USE_EXISTING_MASTER=true kube-up.sh
```

**Comment by druidsbane** [comment]
It looks like MASTER_RESERVED_IP is the issue.  That works for setting the master IP, but when running the kube-up for AWS that isn't checked.  If I export: KUBE_MASTER_IP with the same value as above it now works.  Has that configuration option changed recently for AWS or is this something that needs fixing in the scripts or documentation?  Thanks!

**Pull request merged**